SSA/Projects/Scf/Scf.lean

```diff
@@ -0,0 +1,456 @@
+variable (rgn : Com Op [Ty.int, t] t)
+/- region semantics does not depend on trip count -/
+variable (hrgn : ∀ (v : ⟦t⟧) (n m : Int),
+  Com.denote rgn (Ctxt.Valuation.ofPair n v) =
+  Com.denote rgn (Ctxt.Valuation.ofPair m v))
+
+def lhs : Com Op [/- start-/ .int, /- delta -/.int, /- steps -/ .nat, /- val -/ t] t :=
+  /- v-/
+  /- v1 = -/ Com.lete (for_ (t := t)
+                        ⟨/- start -/ 0, by simp[Ctxt.snoc]⟩
+                        ⟨/- delta -/1, by simp[Ctxt.snoc]⟩
+                        ⟨/- steps -/ 2, by simp[Ctxt.snoc]⟩
+                        ⟨/- v0 -/ 3, by simp[Ctxt.snoc]⟩  rgn) <|
+  Com.ret ⟨0, by simp[Ctxt.snoc]⟩
+
+def rhs : Com Op [/- start-/ .int, /- delta -/.int, /- steps -/ .nat, /- v0 -/ t] t :=
+  /- delta * steps + start-/
+  Com.lete (axpy ⟨1, by simp[Ctxt.snoc]⟩ ⟨2, by simp[Ctxt.snoc]⟩ ⟨0, by simp[Ctxt.snoc]⟩) <|
+  /- -delta -/
+  Com.lete (neg ⟨2, by simp[Ctxt.snoc]⟩) <|
+  Com.lete (for_ (t := t)
+                        ⟨/- end -/ 2, by simp[Ctxt.snoc]⟩
+                        ⟨/- -delta -/ 3, by simp[Ctxt.snoc]⟩
+                        ⟨/- steps -/ 4, by simp[Ctxt.snoc]⟩
+                        ⟨/- v0 -/ 5, by simp[Ctxt.snoc]⟩  rgn) <|
+  Com.ret ⟨0, by simp[Ctxt.snoc]⟩
+
+
+/-- rewrite a variable whose index is '> 0' to a new variable which is the 'snoc' of a smaller variable.
+  this enables rewriting with `Ctxt.Valuation.snoc_toSnoc`. -/
+theorem Ctxt.Var.toSnoc (ty snocty : Ty) (Γ : Ctxt Ty)  (V : Ctxt.Valuation Γ) {snocval : ⟦snocty⟧}
+  {v: ℕ}
+  {hvproof : Ctxt.get? Γ v = some ty}
+  {var : Γ.Var ty}
+  (hvar : var = ⟨v, hvproof⟩) :
+  V var = (Ctxt.Valuation.snoc V snocval) ⟨v+1, by simp[Ctxt.snoc] at hvproof; simp[Ctxt.snoc, hvproof]⟩ := by
+    simp[Ctxt.Valuation.snoc, hvar]
+
+theorem correct :
+  Com.denote (lhs rgn) Γv = Com.denote (rhs rgn) Γv := by
+    simp[lhs, rhs, for_, axpy]
+    try simp (config := {decide := false}) only [
+    Com.denote, Expr.denote, HVector.denote, Var.zero_eq_last, Var.succ_eq_toSnoc,
+    Ctxt.empty, Ctxt.empty_eq, Ctxt.snoc, Ctxt.Valuation.nil, Ctxt.Valuation.snoc_last,
+    Ctxt.ofList, Ctxt.Valuation.snoc_toSnoc,
+    HVector.map, HVector.toPair, HVector.toTuple, OpDenote.denote, Expr.op_mk, Expr.args_mk]
+    generalize A:Γv { val := 0, property := _ } = a;
+    generalize B:Γv { val := 1, property := _ } = b;
+    generalize C:Γv { val := 2, property := _ } = c;
+    generalize D:Γv { val := 3, property := _ } = d;
+    /- we still have things like
+    Ctxt.Valuation.snoc Γv (b * ↑c + a) { val := 2, property := rhs.proof_1 }
+    in the proof state, because it's not sure how to simplify stuff with `snoc`? -/
+    simp[Goedel.toType] at a b c d
+    have h1 : Ctxt.Valuation.snoc (t := .int) Γv (b * (c : ℤ) + a) { val := 2, property := rhs.proof_1 } = b := by
+      conv => rhs; rw[← B]
+    simp[h1]
+    have h2 : Ctxt.Valuation.snoc Γv (t := .int) (b * ↑c + a) { val := 3, property := rhs.proof_2 } = c := by
+      conv => rhs; rw[← C]
+    sorry
```

Comment:
What about this sorry?

Reply:
That's a different proofs, which I will work on next. Note that the proof of the loop transformation in the intro does not use `sorry` (from the `#print axioms correct`, which **does not list** `sorryAx`)